### ISSUE TITLE
Can't change/establish root binding

### ISSUE THREAD

**Issue by TerjeNorderhaug** [thheller/shadow-cljs#520]
Calling `shadow.cljs.devtools.api/nrepl-select` results in a warning:

> IllegalStateException Can't change/establish root binding of: *cljs-compiler-env* with set clojure.lang.Var.set

The source for `nrepl-select` in [shadow.cljs.devtools.api](https://github.com/thheller/shadow-cljs/blob/master/src/main/shadow/cljs/devtools/api.clj) includes the comment:

> ;; for reasons I do not understand this set! fails with
> ;; java.lang.IllegalStateException: Can't change/establish root binding of: *cljs-compiler-env* with set
> ;; although I do check isBound above?

Possibly the fix is to instead use [thread-bound?](https://clojuredocs.org/clojure.core/thread-bound_q) to test if the dynamic var has a thread-local binding, ensuring that set!'ing the var will succeed.

**Comment by thheller** [comment]
This is caused by not having the proper nREPL middleware loaded. I'll see about `thread-bound?` but all that would do is get rid of the error, the middleware is still required to make everything work.

See https://shadow-cljs.github.io/docs/UsersGuide.html#_embedded_nrepl_server

Might need to use the `shadow.cljs.devtools.server.nrepl04` namspace, depends on your nREPL server version (v0.4+).

**Comment by TerjeNorderhaug** [comment]
You are of course right in that it's just bandage to test whether the var is thread-bound before attempting to set it. However, failing to bind cljs-compiler-env doesn't seem to have an impact in my use case (proto-repl on Atom).

I've found that `shadow.cljs.devtools.server.nrepl04/cljs-select` is required `:nrepl-middleware`; without it `shadow.cljs.devtools.api/nrepl-select` will fail with "Can't change/establish root binding of: *nrepl-cljs*".  Yet, even adding some or all of the below `:repl-options` and explicitly have `[nrepl "0.6.0"]` as dependency didn't eliminate the "Can't change/establish root binding of: cljs-compiler-env" warning for `nrepl-select`. Am I missing something?

```clojure
{:nrepl-middleware  
  [shadow.cljs.devtools.server.nrepl04/shadow-init
   shadow.cljs.devtools.server.nrepl04/cljs-load-file
   shadow.cljs.devtools.server.nrepl04/cljs-eval
   shadow.cljs.devtools.server.nrepl04/cljs-select
   cider.piggieback/wrap-cljs-repl]}
```

**Comment by thheller** [comment]
I don't know enough about your setup. The default nrepl server shadow-cljs provides should work just fine and require no additional configuration. If you create your own setup you are pretty much on your own. The middleware list looks correct. Are you sure it is getting loaded properly?

**Comment by TerjeNorderhaug** [comment]
Apparently [shadow.cljs.devtools.api/nrepl-select](https://github.com/thheller/shadow-cljs/blob/master/src/main/shadow/cljs/devtools/api.clj) attempts to set the value of both `cemerick.piggieback/*cljs-compiler-env*` and `cider.piggieback/*cljs-compiler-env*` and **will fail unless both are thread-bound**.

If **both** `cemerick.piggieback/wrap-cljs-repl` and `cider.piggieback/wrap-cljs-repl` are included as middleware the "IllegalStateException" is no longer triggered:

```clojure
:nrepl-middleware
  [shadow.cljs.devtools.server.nrepl04/cljs-select
   cemerick.piggieback/wrap-cljs-repl
   cider.piggieback/wrap-cljs-repl]
```

The reason is that both cljs-compiler-env vars will be bound. They are defined in `shadow.cljs.devtools.server.fake-piggieback` module, required by `shadow.cljs.devtools.server.nrepl04`. 

Without piggieback middleware the two `*cljs-compiler-env*` vars will thus still be bound, so the exception is triggered even then.

If fake-piggieback is required even with no piggieback middleware present, **it seems like `nrepl-select` indeed should NOT warn about the two `cljs-compiler-env` vars not being thread-bound.**

**Comment by thheller** [comment]
Can you please provide you full configuration? I'm still guessing what you are actually doing.

**Comment by TerjeNorderhaug** [comment]
To replicate the bug in `nrepl-select`, create a minimal project with these two files only:

**project.clj**
```clojure
(defproject nrepl-select-bug "0.1.0-SNAPSHOT"                                   
  :dependencies [[thheller/shadow-cljs "2.8.39"]]                               
  :repl-options {:nrepl-middleware                                              
                 [shadow.cljs.devtools.server.nrepl04/cljs-load-file            
                  shadow.cljs.devtools.server.nrepl04/cljs-eval                 
                  shadow.cljs.devtools.server.nrepl04/cljs-select               
                  #_cemerick.piggieback/wrap-cljs-repl                          
                  #_cider.piggieback/wrap-cljs-repl]})                          
```
**shadow-cljs.edn**
```clojure
{:lein true                                                                     
 :builds {:app                                                                  
          {:target :browser                                                     
           :modules {:main {:entries []}}                                       
           :devtools                                                            
           {:http-root   "public"                                               
            :http-port   3000}}}}
```
In `lein repl` evaluate:
```clojure
(require '[shadow.cljs.devtools.server :as server])
(require '[shadow.cljs.devtools.api :as shadow])
(server/start!)
(shadow/watch :app)
(shadow/nrepl-select :app)
```

Two warnings are reported, as expected from the reported bug in [nrepl-select](https://github.com/thheller/shadow-cljs/blob/master/src/main/shadow/cljs/devtools/api.clj):
> :shadow.cljs.devtools.api/piggieback-cemerick
IllegalStateException Can't change/establish root binding of: *cljs-compiler-env* with set
> :shadow.cljs.devtools.api/piggieback-cider
IllegalStateException Can't change/establish root binding of: *cljs-compiler-env* with set

Including **both** the commented-out fake-piggieback middlewares in project.clj eliminates the warnings. I'd expect neither to be required.

Do you want a PR for the fix? It's just replacing the two calls to `.bound` with `thread-bound?` in nrepl-select.

**Comment by thheller** [comment]
Should be fixed in `2.8.40`. Thanks.

**Comment by TerjeNorderhaug** [comment]
Thank you for quickly adding the fix and for all the other work you do on shadow-cljs!

PS: The comment in `nrepl-select` is moot with this change and should perhaps be eliminated. The reason why `set!` fails should now be well understood and it should no longer fail:

> ;; for reasons I do not understand this set! fails with
> ;; java.lang.IllegalStateException: Can't change/establish root binding of: cljs-compiler-env with set
> ;; although I do check isBound above?
